pkg/util/iterutil/iterutil.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2017 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package iterutil
+
+// State iterates over the values.
+//
+// State can be used to create State methods with closures:
+//
+// 	var s iterutil.State
+// 	for _, thing := range myThings {
+// 		s.Update(thing)
+// 		if err := f(&s); err != nil {
+// 			return err
+// 		}
+// 		if s.Done() {
+// 			break
+// 		}
+// 	}
+//
+// where `f` is something like:
+//
+// 	f := func(s *iterutil.State) error {
+// 		repl := s.Cur().(*my.Thing)
+// 		if something(repl) {
+// 			return errors.New("something is not good!")
+// 		} else if somethingElse(repl) {
+// 			return s.Stop() // that's it, won't be called again, even if Stop returns non-nil error
+// 		}
+// 		return nil
+// 	}
+//
+type State struct {
+	cur  interface{}
+	done bool
+}
+
+// Cur returns the current element of the iteration.
+func (i *State) Cur() interface{} { return i.cur }
```

Comment:
Looking at this again, I would make `Cur()` return a

```go
type Cur struct {
  Elem interface{}
  done *bool // is &state.done
}

func (c Cur) Stop() error {
  *c.done = true
  return nil
}
```

and pass `Cur()` to the visitor. That way, the visitor isn't handed something that has an `Update` or `Stopped` method.

Reply:
Yes about this, what I was thinking was that we make `State` an interface like:

```go
type State interface {
	Cur() interface{}
	Stop() error
}
```

Now all iterators can implement `State` as per their convenience. And we can provide one of the implementations (as implemented right now) in the `iterutil` package.

Basically this way is preferred in cases when the `Stop` method does more than just setting `done` to `true`. Not sure if it's required though... If not, your suggestion is the way to go.

Reply:
If `State` is the interface, what is the struct? Ultimately, you will have "two things". In my suggestion, the loop gets `State` and the closure gets `Cur`. In your suggestion, the closure gets `Cur` and the loop... the state impl? That's fine too, as long as the names are sorted out. Maybe you will end up renaming your interface `State` to `Cur` and basically land where I was ;-)

Reply:
Yeah you're right! I'll go with the way you're suggesting. Thanks :D